File: core/types/source.ts
import type { LanguageModel } from "ai";

/**
 * Source type contracts for content ingestion and indexing lifecycle.
 */
/** Metadata extracted or generated during indexing. */
export interface Metadata {
  /** Human-readable title for the source. */
  title: string;
  /** Short description or summary. */
  description: string;
  /** Keypoints derived from the content. */
  keypoints: string[];
}

/**
 * Base class for all indexed sources.
 */
export abstract class IndexedSourceBase {
  /**
   * @param id Unique source ID
   * @param metadata Optional metadata; missing metadata means source is draft.
   */
  constructor(
    public id: string,
    public metadata?: Metadata,
  ) {}

  /** True when source has not been indexed yet. */
  get isDraft(): boolean {
    return !this.metadata;
  }

  /** Source kind discriminator for indexed sources. */
  abstract readonly kind: string;
  /** True when content should be consumed as a list/stream. */
  abstract get isIterable(): boolean;
  /** Returns content or a lazy/async provider. */
  abstract getContent: () =>
    | string
    | string[]
    | AsyncIterable<string>
    | Promise<string | string[] | AsyncIterable<string>>;

  /** Build the input string used for indexing. */
  abstract buildIndexInput(): string | Promise<string>;
  /** Build the input string used for deep judging. */
  abstract buildDeepJudgeInput(): string | Promise<string>;
}

File: python/src/contextrie/types/source.py
from __future__ import annotations

from abc import ABC, abstractmethod
from collections.abc import AsyncIterable
from dataclasses import dataclass
from typing import Any


@dataclass(slots=True)
class Metadata:
    """Metadata extracted or generated during indexing."""

    title: str
    description: str
    keypoints: list[str]

    def __post_init__(self) -> None:
        if not isinstance(self.title, str) or not self.title.strip():
            raise ValueError("Metadata.title must be a non-empty string")
        if not isinstance(self.description, str) or not self.description.strip():
            raise ValueError("Metadata.description must be a non-empty string")
        if not isinstance(self.keypoints, list) or not self.keypoints:
            raise ValueError("Metadata.keypoints must be a non-empty list")
        if not all(isinstance(point, str) and point.strip() for point in self.keypoints):
            raise ValueError("Metadata.keypoints must contain non-empty strings")


class IndexedSourceBase(ABC):
    """Base class for retrieval-ready sources."""

    kind: str

    def __init__(
        self,
        id: str,
        metadata: Metadata | None = None,
        path: str | None = None,
    ) -> None:
        self.id = id
        self.metadata = metadata
        self.path = path

    @property
    def is_draft(self) -> bool:
        return self.metadata is None

    @property
    @abstractmethod
    def is_iterable(self) -> bool:
        raise NotImplementedError

    @abstractmethod
    def get_content(self) -> str | list[str] | AsyncIterable[str] | Any:
        raise NotImplementedError

    @abstractmethod
    async def build_index_input(self) -> str:
        raise NotImplementedError

    @abstractmethod
    async def build_deep_judge_input(self) -> str:
        raise NotImplementedError
